server/src/main/java/io/crate/expression/scalar/DateBinFunction.java

```diff
@@ -0,0 +1,215 @@
+
+public class DateBinFunction extends Scalar<Long, Object> {
+
+    public static final String NAME = "date_bin";
+
+    private final Function<Object, Long> intervalOperator;
+
+    // First argument is a result of intervalOperator
+    private final BiFunction<Long, Object, Long> timeBucketOperator;
+
+    // Interval is same for all rows, computed once in compile()
+    // in order to avoid toStandardDuration() call and object allocation per row.
+    @Nullable
+    private Long compiledInterval;
+
+    private enum IntervalOperators {
+        LONG(
+            (Function<Object, Long>) interval -> (Long) interval
+        ),
+        INTERVAL(
+            (Function<Object, Long>) interval -> Long.valueOf(((Period) interval).toStandardDuration().getMillis())
+        );
+
+        private final Function function;
+
+        IntervalOperators(Function function) {
+            this.function = function;
+        }
+
+        public Function getFunction() {
+            return function;
+        }
+    }
+
+    private enum BucketOperators {
+        LONG(
+            (BiFunction<Long, Object, Long>) (interval, timestamp) -> getBucketedTimestamp(interval, (long) timestamp)
+        ),
+        TIMESTAMP(
+            (BiFunction<Long, Object, Long>) (interval, timestamp) -> getBucketedTimestamp(interval, DataTypes.TIMESTAMP.implicitCast(timestamp))
+        ),
+        TIMESTAMPZ(
+            (BiFunction<Long, Object, Long>) (interval, timestamp) -> getBucketedTimestamp(interval, DataTypes.TIMESTAMPZ.implicitCast(timestamp))
+        );
+
+
+        private static long getBucketedTimestamp(long interval, long timestamp) {
+            if (interval <= 0) {
+                throw new IllegalArgumentException("Interval cannot be negative or equal to zero");
+            }
+
+            /*
+             in Java % operator returns negative result only if dividend is negative.
+             https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.17.3
+
+             We need to shift timestamp by the timeline to the "earlier" direction, to the left.
+             But if timestamp is negative (before 1970), remainder will be also negative (see link above)
+             and we need to add it as subtracting would move it to the opposite direction.
+             See test_interval_timestamp_without_zone_works as example.
+            */
+
+            if (timestamp < 0) {
+                return timestamp - timestamp % interval - interval;
+            }
+            return timestamp - timestamp % interval;
+        }
+
+        private final BiFunction function;
+
+        BucketOperators(BiFunction function) {
+            this.function = function;
+        }
+
+        public BiFunction getFunction() {
+            return function;
+        }
+    }
+
+    public static void register(ScalarFunctionModule module) {
+
+        List<DataType<?>> supportedIntervalTypes = List.of(
+            DataTypes.LONG, DataTypes.INTERVAL);
```

Comment:
Can we restrict this to only `INTERVAL`? I don't think we need to support `LONG`. I think it might even work with number due to implicit casts, and if it doesn't it is probably better anyway because it would force users to add explicit casts - which would improve readability by them having to clarify the unit of the numbers.

Reply:
👌 I will drop Long in the next commit.